pairing/bn256/point.go

```diff
@@ -62,21 +62,65 @@ func (p *pointG1) Clone() kyber.Point {
 }
 
 func (p *pointG1) EmbedLen() int {
-	panic("bn256.G1: unsupported operation")
+	// 2^255 is ~size of the curve P
+	// minus one byte for randomness
+	// minus one byte for len(data)
+	return (255 - 8 - 8) / 8
 }
 
 func (p *pointG1) Embed(data []byte, rand cipher.Stream) kyber.Point {
-	// XXX: An approach to implement this is:
-	// - Encode data as the x-coordinate of a point on y²=x³+3 where len(data)
-	//   is stored in the least significant byte of x and the rest is being
-	//   filled with random values, i.e., x = rand || data || len(data).
-	// - Use the Tonelli-Shanks algorithm to compute the y-coordinate.
-	// - Convert the new point to Jacobian coordinates and set it as p.
-	panic("bn256.G1: unsupported operation")
+	// How many bytes to embed?
+	dl := p.EmbedLen()
+	if dl > len(data) {
+		dl = len(data)
+	}
+
+	intCurveB := curveB.BigInt()
+
+	for {
+		// Pick a random point, with optional embedded data
+		var b [32]byte
+		rand.XORKeyStream(b[:], b[:])
+		if data != nil {
+			b[0] = byte(dl)       // Encode length in low 8 bits
+			copy(b[1:1+dl], data) // Copy in data to embed
+		}
+		x := new(big.Int).SetBytes(b[:])
+
+		xxx := new(big.Int).Mul(x, x)
+		xxx.Mul(xxx, x)
+		xxx.Mod(xxx, P)
```

Comment:
I would prefer to keep that prime unchanged, also you can see that this piece of code is similar to `hashToPoint` (well it's doing the same thing basically). One simple trick to avoid the prime `p` to change is to declare a function `deriveY`:

```go
// deriveY compute the y coordinate of x for x^3 + b = y^2
func deriveY(x *big.Int, curveB *big.Int) (*big.Int) {
  xxx := new(big.Int).Mul(x, x)
  xxx.Mul(xxx, x)
  xxx.Mod(xxx, p)

  t := new(big.Int).Add(xxx, intCurveB)
  return new(big.Int).ModSqrt(t, p)
}
```

You can then hook that up in your Embed and in the hashToPoint.

Reply:
You da boss! I thought about doing it like this but decided against it for some reason. Should be in now